### PR TITLE
Add async csv

### PR DESCRIPTION
**Pull request by Radiergummi**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present)

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration` (**Note: Not in the currently published version on NPM, but on master**)
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

----

This adds types for the [`async-csv`](https://www.npmjs.com/package/async-csv) package. While the maintainer has started working on a new version with typescript support, requests to finish or publish the current state have not received an answer for a while, so it makes sense to publish types for the current stable package.  
The types are in parts copied from the existing types in the repository; I tested them in my own project to verify they work with the currently released code.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## async-csv ([<kbd>unpkg</kbd>](https://unpkg.com/browse/async-csv@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'async-csv'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 541ae6d1349c93fb87d8332dbc6a6665c146cb71
</p>

**Comment by Radiergummi** [comment]
> Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
> The check for missing properties isn't always right, so take this list as advice, not a requirement.
> 
> ## async-csv ([unpkg](https://unpkg.com/browse/async-csv@latest/))
> was missing the following properties:
> 
> 1. The declaration doesn't match the JavaScript module 'async-csv'. Reason:
>    The JavaScript module can be called or constructed, but the declaration module cannot.
> 
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [541ae6d](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/541ae6d1349c93fb87d8332dbc6a6665c146cb71)

The Javascript module simply exports a class with four static methods, making it effectively the same as an object with four method properties.

**Comment by peterblazejewicz** [comment]
> This adds types for the [`async-csv`](https://www.npmjs.com/package/async-csv) package. While the maintainer has started working on a new version with typescript support, requests to finish or publish the current state have not received an answer for a while, so it makes sense to publish types for the current stable package.

Can you pls reference that thread?
the `package.json` just misses the export for typings (`typings`/'files'/,etc), so let's try to avoid duplication in that area, thx!

**Comment by Radiergummi** [comment]
The thread is [issue 4](https://github.com/anton-bot/async-csv/issues/4) on the repository.

**Comment by peterblazejewicz** [comment]
cool, I can see. Sorry for asking those questions.

So if that is to be consumed like the CJS, what about:
```ts
// Type definitions for async-csv 2.1
// Project: https://github.com/anton-bot/async-csv#readme
// Definitions by: Moritz Friedrich <https://github.com/Radiergummi>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

/// <reference types="node" />
declare const CsvAsync: {
    /**
     * @param options
     */
    generate: (options?: CsvAsync.CsvGenerateOptions) => Promise<unknown>;

    /**
     * Parses a CSV file into an array of rows.
     *
     * @param input
     * @param options
     */
    parse: (input: string, options?: CsvAsync.CsvParseOptions) => Promise<unknown>;

    /**
     *
     * @param data
     * @param handler
     * @param options
     */
    transform: <T = any, U = any>(
        data: T[],
        handler: (record: T, callback: (err?: Error | null, record?: T) => void, params?: any) => U,
        options?: CsvAsync.TransformOptions,
    ) => Promise<U[]>;

    /**
     * Converts an array of rows into a CSV string.
     *
     * @param data
     * @param options
     */
    stringify: (
        data: Array<Array<string | number | null | undefined>>,
        options?: CsvAsync.CsvStringifyOptions,
    ) => Promise<string>;
};

declare namespace CsvAsync {
    interface CsvGenerateOptions {
        ...
		[...] there is more here
        ...
    }
}

export = CsvAsync;
```
(and then we can skip):
```json
{
    "extends": "dtslint/dt.json"
}
```
all your test code can be left as is.
Is this how it is consumed with current package version?

**Comment by Radiergummi** [comment]
Oh, yeah that's awesome. Didn't know you could have a const and namespace with the same name, that makes the export situation way better 😊 Thank you!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@Radiergummi thx for the contribution!
Maybe someone from your community review this as well, thx!

File: types/async-csv/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+     */
+    generate(options?: CsvAsync.CsvGenerateOptions): Promise<unknown>;
+
+    /**
+     * Parses a CSV file into an array of rows.
+     *
+     * @param input
+     * @param options
+     */
+    parse(input: string, options?: CsvAsync.CsvParseOptions): Promise<unknown>;
```
Comment (Flouwrian) [inline types/async-csv/index.d.ts:20]: du kannst den type PRomise<unknown> zu Promise<void> machen, wenn das ding keinen rückgabewert hat. Falls schon würde ich den entweder zusätzlichtypen oder Promise<any> nehmen.

**Review by Flouwrian — APPROVED** [review]
traum modul <3

**Review by peterblazejewicz — APPROVED** [review]

**Comment by orta** [comment]
:+1:

**Pull request merged**